pkg/stanza/fileconsumer/internal/tracker/tracker.go

```diff
@@ -144,6 +155,26 @@ func (t *fileTracker) TotalReaders() int {
 	return total
 }
 
+func (t *fileTracker) restoreArchiveIndex() {
+	if !t.archiveEnabled() {
+		return
+	}
+	byteIndex, err := t.persister.Get(context.Background(), archiveIndexKey)
+	if err != nil {
+		t.set.Logger.Error("error while reading the archiveIndexKey. Starting from 0", zap.Error(err))
+		return
+	}
+	t.archiveIndex, err = byteToIndex(byteIndex)
+	if err != nil {
+		t.set.Logger.Error("error getting read index. Starting from 0", zap.Error(err))
+	} else if t.archiveIndex < 0 || t.archiveIndex >= t.pollsToArchive {
+		// safety check. It can happen if `polls_to_archive` was changed.
+		// It's best if we reset the index or else we might end up writing invalid keys
+		t.set.Logger.Warn("the read index was found, but it exceeds the bounds. Starting from 0")
+		t.archiveIndex = 0
+	}
```

Comment:
Good idea to check for this case.

However, I wonder if we can handle it better than restarting from zero. What would it take to search the archive for the most recently updated?

I think we could maintain some kind of data structure which notes the time each archive was written. Maybe just `map[index]time.Time`. Then when we first create the tracker, we can load this up and find the most recent timestamp. We can also check for the case where `pollsToArchive` has changed and then rewrite the storage to align with the new value. 

For example, if we previously saved 10 archives and find that `pollsToArchive` is now 5, we can find the 5 most recent indices based on the timestamp structure, then rewrite the archive files so that these are 0-4. We should probably even delete the extras from storage as well.

Reply:
This solution does makes sense to me, but it becomes tricky when we eventually overwrite old archive data, as it is a ring buffer.
We might need to load the filesets in memory. 
I'll find a few ways.

Reply:
> it becomes tricky when we eventually overwrite old archive data, as it is a ring buffer.

Can you elaborate?

> We might need to load the filesets in memory.

If it's more than one at a time then it defeats the point of the archive.

Reply:
> Can you elaborate?

Consider this archive,

<img width="511" alt="Screenshot 2024-12-23 at 8 37 38 PM" src="https://github.com/user-attachments/assets/4e5d8bec-fec1-4422-89bf-81ac1827853d" />

We've rolled over once and the latest data is at index `4` and `archiveIndex` (i.e. where the next data will be written) is at index `5`. 

Let's suppose that new `polls_to_archive` is 7.
We now need to construct a new, smaller archive with 7 most recent elements. 
These elements are (from most recent to least recent):

```14, 13, 12, 11, 10, 9, 8```

We cannot simply rewrite archive in-place without caching values. 

It would be much simpler to convert archive like following image,

<img width="572" alt="Screenshot 2024-12-23 at 8 41 43 PM" src="https://github.com/user-attachments/assets/e5492c97-992c-491d-b511-16b46fbd5aa1" />

and we would delete excess data. 

Wdyt?

Reply:
> What would it take to search the archive for the most recently updated?

It would always be data stored at `archiveIndex-1` index. We will store `archiveIndex` on disk, so in next collector run, we would load that value and we can find most recent data.

`archiveIndex` points at the next location where data will be written. 
This can point to either of following:
- Least recent data
- Pointing to an empty slot (archive is partially filled)